sim/aviron/src/main.zig

```diff
@@ -85,21 +95,33 @@ pub fn main() !u8 {
             .elf => {
                 var header = try std.elf.Header.read(&reader.interface);
 
+                // Set PC to entry point (convert byte address to word address for AVR)
+                cpu.pc = @intCast(header.entry / 2);
+
                 var pheaders = header.iterateProgramHeaders(&reader);
                 while (try pheaders.next()) |phdr| {
                     if (phdr.p_type != std.elf.PT_LOAD)
-                        continue; // Header isn't lodead
+                        continue; // Header isn't loaded
 
-                    const dest_mem = if (phdr.p_paddr >= 0x0080_0000)
+                    if (phdr.p_memsz == 0)
+                        continue; // Empty segment, nothing to load
+
+                    // Use vaddr to determine if this is data or code
+                    // AVR uses 0x800000 flag in vaddr to indicate data memory
+                    var dest_mem: []u8 = if (phdr.p_vaddr >= 0x0080_0000)
                         &sram.data
                     else
                         &flash_storage.data;
 
-                    const addr_masked: u24 = @intCast(phdr.p_paddr & 0x007F_FFFF);
+                    const addr_masked: u24 = @intCast(phdr.p_vaddr & 0x007F_FFFF);
+                    const target_addr: u24 = if (phdr.p_vaddr >= 0x0080_0000)
+                        addr_masked - @as(u24, mcu_config.sram_base)
```

Comment:
The subtraction of `mcu_config.sram_base` is still wrong here, as `0x0080_0000` is the start of the *data* bus, not the *sram*. This would always load the memory.

I would fully go through the bus map here and load the data with `databus.write()` byte by byte, only using shortcuts if always applicable.

Reply:
Fixed